### ISSUE TITLE
Finish PHP extenders

### ISSUE THREAD

**Issue by franzliedke** [flarum/core#1891]
Since [introducing the extender API in beta.8](https://github.com/flarum/core/issues/851), a bit of time has passed and we still need to add extenders for the following use-cases.

This list should cover what we need to implement all official, bundled extensions' backends only using existing extenders.

## Remaining Post-Beta 15 Steps:

- [x] Finish up and merge Search extenders
- [x] Review and merge user password checker extenders
- [x] Look over extender naming (perhaps mutate => attributes for ApiSerializer?)
- [x] Add a 4th argument default value to Setting extender
- [x] Add a beforeSending callback method to Notification extender
- [x] Figure out what to do with GetModelIsPrivate
- [x] Add `subscribe` to event extender, but register subscribers only after application is fully booted to avoid early resolution.

## Extenders To Replace Events

At this stage, our goal is to replace the current event-based extension system for non-domain events, not to introduce new extensibility functionality (except where we can break current ambiguous events into more granular, use-driven extenders).

- [x] **Extend\ApiSchema**  (#1624)
  - [x] Api\Event\Serializing
  - [x] Api\Event\WillGetData
  - [x] Api\Event\WillSerializeData
  - [x] Event\GetApiRelationship

- [x] **Extend\Console**
  - [x] Console\Event\Configuring

- [x] **Extend\Formatter** (#2098) @askvortsov1 
  - [x] Formatter\Event\Configuring
  - [x] Formatter\Event\Parsing
  - [x] Formatter\Event\Rendering

- [x] **Extend\Floodgate** (#2170) @askvortsov1 
  - [x] Post\Event\CheckingForFlooding

- [x] **Extend\User**
  - [x] User\Event\GetDisplayName
  - [x] Event\ConfigureUserPreferences (#2005)
  - [x] User\Event\CheckingPassword (#2176) @askvortsov1 
  - [x] User\Event\PrepareUserGroups

- [x] **Extend\Policy**
  - [x] User\Event\GetPermission (#2056) @askvortsov1 
  - [x] User\Event\ScopeModelVisibility @askvortsov1 

- [x] **Extend\Search** (#2104) @askvortsov1 
  - [x] Event\AbstractConfigureGambits
  - [x] Event\ConfigureDiscussionGambits
  - [x] Event\ConfigureUserGambits
  - [x] Discussion\Event\Searching
  - [x] User\Event\Searching
  - [x] Event\ConfigurePostsQuery

- [x] **Extend\Routes**
  - [x] Event\AbstractConfigureRoutes
  - [x] Event\ConfigureApiRoutes
  - [x] Event\ConfigureForumRoutes

- [x] **Extend\Locales**
  - [x] Event\ConfigureLocales

- [x] **Extend\Middleware**
  - [x] Event\ConfigureMiddleware

- [x] **Extend\Model**
  - [x] Event\ConfigureModelDates
  - [x] Event\ConfigureModelDefaultAttributes
  - [x] Event\GetModelRelationship

- [x] **Extend\Notification**
  - [x] Event\ConfigureNotificationTypes @SychO9 
  - [x] Notification\Event\Sending

- [x] **Extend\Post**
  - [x] Event\ConfigurePostTypes

- [x] **Extend\Validation** (#2102) @askvortsov1 
  - [x] Foundation\Event\Validating

## Events to Drop

- Event\GetModelIsPrivate (The Saving event for Post / Discussion gets the job done more simply)

## Events To Keep

**Discussion**
- Discussion\Event\Deleted
- Discussion\Event\Deleting
- Discusion\Event\Hidden
- Discussion\Event\Renamed
- Discussion\Event\Restored
- Discussion\Event\Saving
- Discussion\Event\Started
- Discussion\Event\UserDataSaving
- Discussion\Event\UserRead

**Extension**
- Extension\Event\Disabled
- Extension\Event\Disabling
- Extension\Event\Enabled
- Extension\Event\Enabling
- Extension\Event\Uninstalled

**Group**
- Group\Event\Created
- Group\Event\Deleted
- Group\Event\Deleting
- Group\Event\Renamed
- Group\Event\Saving

**Post**
- Post\Event\Deleting
- Post\Event\Deleted
- Post\Event\Hidden
- Post\Event\Restored
- Post\Event\Revised
- Post\Event\Saving

**Settings**
- Settings\Event\Saved
- Settings\Event\Saving
- Settings\Event\Deserializing
- Settings\Event\Serializing

**User**
- User\Event\Activated
- User\Event\AvatarChanged
- User\Event\AvatarDeleting
- User\Event\AvatarSaving
- User\Event\BioChanged
- User\Event\Deleted
- User\Event\Deleting
- User\Event\EmailChanged
- User\Event\EmailChangeRequested
- User\Event\GroupsChanged
- User\Event\LoggedIn
- User\Event\LoggedOut
- User\Event\Registered
- User\Event\RegisteringFromProvider
- User\Event\Renamed
- User\Event\Saving

## OLD LIST

- [x] API: Schema (Attributes / Relationships / Serialize)
- [x] API: load data
- [x] Console: Commands (#2057)
- [x] Core: Error handlers (#1781)
- [x] CSRF protection: Whitelist routes (#1941, #2095)
- [x] Event listeners for domain events (#2097)
- [x] Formatter: rendering, parsing (#2098)
- [x] Middleware: registration (#1919)
- [x] Mail: drivers (#2012)
- [x] Model: default attributes (#2100)
- [x] Model: relationship (#2100)
- [x] Model: date attributes (#2100)
- [x] Notification: Sending
- [x] Notification: types
- [x] Policies: e.g. get permission on model
- [x] Posts: types (#2101)
- [x] Search: reorder (custom logic)
- [x] Search: gambit (#2104)
- [x] User: preferences
- [x] User: saving
- [x] User: prepare user groups (#2110)
- [x] User: display name (#2174)
- [x] Validator: configuration (#2102)
- [x] View: Namespace (#2134)

Most of them come from [Toby's original gist](https://gist.github.com/tobyzerner/5e514a700dcb471180369903ab4562c3).

Other related issues: #1223

**Comment by luceos** [comment]
Looking at your list I am unsure how you want to move forward. It feels like adding extenders for all of them feels like we're bloating core. We could potentially even make the Events themselves smarter to allow for easier extensibility.

Looking at Toby's gist I see a lot of possible unnecessary duplication, eg:

```php
new Extend\DiscussionGambit(TagGabmit::class),
```

Why not:

```php
new Extend\Gambit(Discussion::class, TagGabmit::class),
```

Maybe I'm completely oblivious to the exact plans.

**Comment by franzliedke** [comment]
> It feels like adding extenders for all of them feels like we're bloating core.

I think the more focused public API and the greater freedom to make internal changes without having to change the API for extensions is absolutely worth it. Once we have the public API, we can get rid of the events as implementation detail in many cases, which should pretty much reduce the bloat by the same amount we increased it.

> We could potentially even make the Events themselves smarter to allow for easier extensibility.

Not sure I understand - can you expand on this? Maybe with an example?

> Looking at Toby's gist I see a lot of possible unnecessary duplication

Yeah. We can probably implement them as one. Unless something comes up that shows why Toby proposed them the way he did. :wink:

**Comment by luceos** [comment]
Here's an example from tenancy where we've added a few helper methods on the event directly to simplify implementation for the developer:

```php
<?php

declare(strict_types=1);

/*
 * This file is part of the tenancy/tenancy package.
 *
 * Copyright Tenancy for Laravel & Daniël Klabbers <daniel@klabbers.email>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @see https://tenancy.dev
 * @see https://github.com/tenancy
 */

namespace Tenancy\Affects\Routes\Events;

use Illuminate\Routing\RouteCollection;
use Illuminate\Routing\Router;
use Tenancy\Identification\Events\Switched;

class ConfigureRoutes
{
    /**
     * @var Switched
     */
    public $event;

    /**
     * @var Router
     */
    public $router;

    public function __construct(Switched $event, Router $router)
    {
        $this->event = $event;
        $this->router = $router;
    }

    /**
     * Flush all tenant routes for this request.
     *
     * @return $this
     */
    public function flush()
    {
        $this->router->setRoutes(new RouteCollection());

        return $this;
    }

    /**
     * Adds routes from a routes.php file to the current request.
     *
     * @param array  $attributes
     * @param string $path
     *
     * @return $this
     */
    public function fromFile(array $attributes, string $path)
    {
        $this->router->group($attributes, $path);

        return $this;
    }
}
```

**Comment by cmcjacob** [comment]
I think there's a heavy boilerplate with the current state of relying so much on event listeners.  They are implemented well, but come with a degree of importance that it can give the impression they  
 are the only option going forward.

Most of the helpers in this list sound beneficial as extenders, but it seems like some of them should be decoupled as officially enabled extensions, so kind of confused how core would be considered bloated from a supplicant feature.

Furthermore, if (hypothetically) 90% of events are left as blank containers and a good 15% are enhanced with app/logic helpers - could provide an unwanted layer of inconsistency that doesn't contrast well with the existing extender methods.

**Comment by franzliedke** [comment]
@cmcjacob Most extenders will rely on the existing events for implementation in the beginning. But as soon as the extenders are there, extension developers will be expected to use those - events will then be considered *private API*.

As you mention, some of the events are quite clunky for some of their use-cases, which is why we will likely change the implementation of extenders to something more direct (e.g. storing an array of middleware per "frontend" in the IoC container directly).

**Comment by franzliedke** [comment]
For my own reference: I noticed that [this GitHub code search](https://github.com/search?q=Flarum%5CExtend%5CExtenderInterface&type=Code) is very useful to find out what extension developers are currently building custom extenders for. :smiley:

**Comment by askvortsov1** [comment]
Please forgive me if this question is a bit uninformed, but is the goal of this to shift the vast majority of extension-flarum interaction to the extender interface (in effect, removing the necessity of listening into events)?

Also, if that's the case, I noticed that there's a lack of extenders that have to do with logging in / registering. There are valid cases for actions to be executed after login (see https://github.com/askvortsov1/flarum-auth-sync), so I think that's something we should plan for.

Essentially, I'm wondering how the extenders to have were selected, and to what extend this list is set in stone?

**Comment by franzliedke** [comment]
> is the goal of this to shift the vast majority of extension-flarum interaction to the extender interface (in effect, removing the necessity of listening into events)?

Yes, pretty much. Whether they use events internally or not, is going to be an implementation detail and considered private API. Basically, the extenders (and what other classes the extenders reference and expose) will make up the public API of Flarum's backend.

> Essentially, I'm wondering how the extenders to have were selected...

Basically, I want all bundled extensions to be implemented fully in terms of extenders. They cover a variety of use-cases (experimenting with those use-cases was one of the reasons why the extensions were implemented so early), so we are hoping to cover enough ground with these. Additional input / requests on top are, of course, welcome, but I'd consider them in nice-to-have category.

> ...and to what extend this list is set in stone?

It's not, see above. Especially the exact interfaces of the concrete extenders are up to discussion - best kept to concrete sub-tickets, though.

Baseline for discussion would be [this Gist by Toby, Flarum's founder](https://gist.github.com/tobyzerner/5e514a700dcb471180369903ab4562c3), where he envisioned what the tags extension would look like if implemented using extenders.

**Comment by franzliedke** [comment]
P.S.: I want extensions to mostly *use* extenders, not build their own. The latter has become a bit too common for my personal taste, but that's mostly due to our slow pace in creating more official extenders. :wink:

**Comment by askvortsov1** [comment]
> It's not, see above. Especially the exact interfaces of the concrete extenders are up to discussion - best kept to concrete sub-tickets, though.

I think that to organize work on this, it might be best to start this discussion and decide which ones we definitely want, as well as any kind of priority we want to do these in.

**Comment by franzliedke** [comment]
We'll organize a call / chat meeting in early March, after beta.12 is released. Everybody who's interested, let us know and we'll give you a chance to join.

**Comment by askvortsov1** [comment]
For the event listener ones, what if we refactored all the created/deleted/saving/etc events out into model-agnostic ones, and created a general use AbstractLifecycleHandler, which extension devs could override and register via Extend\Model->addLifecycleHandler?

The lifecycle handler could have methods ie onDeleting, onDeleted, onCreating, on Created, etc which, if defined by extension devs, would be subscribed onto the new model-agnostic events. 

This should simplify/add flexibility to the extender system, allow the per-model extenders (e.g. the User or Group or Post or Notification extenders) to focus on model-specific situations, and also fix https://github.com/flarum/core/issues/1223.
